File: Easy/0009-palindrome-number/scala/palindrome2.scala
object PalindromeHalfReverse {

  def isPalindrome(x: Int): Boolean = {
    // Negative numbers are not palindromes.
    // Also, numbers ending with 0 are not palindromes unless the number is exactly 0
    // (because a palindrome can't start with 0).
    if (x < 0 || (x % 10 == 0 && x != 0)) return false

    @annotation.tailrec
    def loop(n: Int, revHalf: Int): Boolean = {
      // Stop when we've reversed at least half of the digits.
      if (n <= revHalf) {
        // Even length: n == revHalf
        // Odd length:  n == revHalf / 10 (skip the middle digit)
        n == revHalf || n == revHalf / 10
      } else {
        val lastDigit = n % 10
        loop(n / 10, revHalf * 10 + lastDigit)
      }
    }

    loop(x, 0)
  }
}

File: Easy/0009-palindrome-number/ocaml/palindrome2.ml
let is_palindrome x =
  (* Negative numbers are not palindromes.
     If it ends with 0, it can only be palindrome if it's exactly 0. *)
  if x < 0 || (x mod 10 = 0 && x <> 0) then false
  else
    let rec loop n rev_half =
      (* Stop when we've processed at least half the digits. *)
      if n <= rev_half then
        (n = rev_half) || (n = rev_half / 10)
      else
        let last_digit = n mod 10 in
        loop (n / 10) (rev_half * 10 + last_digit)
    in
    loop x 0
